File: backend/scan_errors.py
#!/usr/bin/env python3
"""
Backend Error Scanner
Scans all Python files for common errors and issues.
"""

import os
import sys
import ast
import subprocess
from pathlib import Path
from typing import List, Dict, Tuple


# Colors for terminal output
class Colors:
    RED = "\033[0;31m"
    GREEN = "\033[0;32m"
    YELLOW = "\033[1;33m"
    BLUE = "\033[0;34m"
    NC = "\033[0m"  # No Color


def print_header(text: str):
    print(f"\n{Colors.BLUE}{'='*60}")
    print(f"{text}")
    print(f"{'='*60}{Colors.NC}\n")


def print_success(text: str):
    print(f"{Colors.GREEN}✓ {text}{Colors.NC}")


def print_error(text: str):
    print(f"{Colors.RED}✗ {text}{Colors.NC}")


def print_warning(text: str):
    print(f"{Colors.YELLOW}⚠ {text}{Colors.NC}")


def scan_syntax_errors(directory: str) -> List[Tuple[str, str]]:
    """Scan for Python syntax errors."""
    print_header("Scanning for Syntax Errors")
    errors = []

    for py_file in Path(directory).rglob("*.py"):
        try:
            with open(py_file, "r", encoding="utf-8") as f:
                ast.parse(f.read(), filename=str(py_file))
        except SyntaxError as e:
            errors.append((str(py_file), f"Line {e.lineno}: {e.msg}"))
            print_error(f"{py_file}: Line {e.lineno}: {e.msg}")

    if not errors:
        print_success("No syntax errors found!")
    else:
        print_error(f"Found {len(errors)} syntax errors")

    return errors


def scan_import_errors(directory: str) -> List[Tuple[str, str]]:
    """Scan for import errors."""
    print_header("Scanning for Import Errors")
    errors = []

    # Change to directory to ensure imports work
    original_dir = os.getcwd()
    os.chdir(directory)

    for py_file in Path(".").rglob("*.py"):
        if "__pycache__" in str(py_file) or ".venv" in str(py_file):
            continue

        try:
            with open(py_file, "r", encoding="utf-8") as f:
                tree = ast.parse(f.read())

            for node in ast.walk(tree):
                if isinstance(node, (ast.Import, ast.ImportFrom)):
                    # Extract module names
                    if isinstance(node, ast.Import):
                        modules = [alias.name for alias in node.names]
                    else:
                        modules = [node.module] if node.module else []

                    for module in modules:
                        if module and not module.startswith("."):
                            try:
                                __import__(module.split(".")[0])
                            except ImportError as e:
                                error_msg = f"Cannot import '{module}': {str(e)}"
                                errors.append((str(py_file), error_msg))
                                print_warning(f"{py_file}: {error_msg}")
        except Exception as e:
            print_warning(f"Could not analyze {py_file}: {e}")

    os.chdir(original_dir)

    if not errors:
        print_success("No import errors found!")
    else:
        print_warning(f"Found {len(errors)} potential import issues")

    return errors


def run_flake8(directory: str) -> bool:
    """Run flake8 linter."""
    print_header("Running Flake8 Linter")

    try:
        result = subprocess.run(
            [
                "flake8",
                directory,
                "--max-line-length=120",
                "--exclude=.venv,__pycache__,.pytest_cache",
            ],
            capture_output=True,
            text=True,
        )

        if result.returncode == 0:
            print_success("No flake8 issues found!")
            return True
        else:
            print_error("Flake8 found issues:")
            print(result.stdout)
            return False
    except FileNotFoundError:
        print_warning("flake8 not installed. Install with: pip install flake8")
        return False


def run_bandit(directory: str) -> bool:
    """Run bandit security scanner."""
    print_header("Running Bandit Security Scanner")

    try:
        result = subprocess.run(
            ["bandit", "-r", directory, "-ll", "--exclude", ".venv,__pycache__"],
            capture_output=True,
            text=True,
        )

        if "No issues identified" in result.stdout:
            print_success("No security issues found!")
            return True
        else:
            print_warning("Bandit found potential security issues:")
            print(result.stdout)
            return False
    except FileNotFoundError:
        print_warning("bandit not installed. Install with: pip install bandit")
        return False


def check_requirements(directory: str) -> bool:
    """Check if all requirements are installed."""
    print_header("Checking Requirements")

    req_file = Path(directory) / "requirements.txt"
    if not req_file.exists():
        print_warning("requirements.txt not found")
        return False

    try:
        result = subprocess.run(["pip", "check"], capture_output=True, text=True)

        if result.returncode == 0:
            print_success("All requirements are satisfied!")
            return True
        else:
            print_error("Dependency issues found:")
            print(result.stdout)
            return False
    except Exception as e:
        print_error(f"Error checking requirements: {e}")
        return False


def generate_report(results: Dict) -> None:
    """Generate summary report."""
    print_header("SCAN SUMMARY")

    total_issues = sum(len(v) if isinstance(v, list) else (0 if v else 1) for v in results.values())

    print(f"Syntax Errors: {len(results['syntax_errors'])}")
    print(f"Import Issues: {len(results['import_errors'])}")
    print(f"Flake8: {'✓ PASS' if results['flake8'] else '✗ FAIL'}")
    print(f"Bandit: {'✓ PASS' if results['bandit'] else '✗ FAIL'}")
    print(f"Requirements: {'✓ PASS' if results['requirements'] else '✗ FAIL'}")

    print(f"\n{Colors.BLUE}Total Issues: {total_issues}{Colors.NC}")

    if total_issues == 0:
        print(f"\n{Colors.GREEN}🎉 All checks passed! Your code looks good!{Colors.NC}")
    else:
        print(f"\n{Colors.YELLOW}⚠️  Please fix the issues above before deploying.{Colors.NC}")


def main():
    # Determine backend directory
    script_dir = Path(__file__).parent
    backend_dir = script_dir / "app"

    if not backend_dir.exists():
        backend_dir = script_dir

    print(f"{Colors.BLUE}Backend Error Scanner{Colors.NC}")
    print(f"Scanning directory: {backend_dir}\n")

    results = {
        "syntax_errors": scan_syntax_errors(str(backend_dir)),
        "import_errors": scan_import_errors(str(backend_dir)),
        "flake8": run_flake8(str(backend_dir)),
        "bandit": run_bandit(str(backend_dir)),
        "requirements": check_requirements(str(script_dir)),
    }

    generate_report(results)

    # Exit with error code if issues found
    total_issues = len(results["syntax_errors"]) + len(results["import_errors"])
    sys.exit(1 if total_issues > 0 else 0)


if __name__ == "__main__":
    main()

File: frontend/web/scan_errors.js
#!/usr/bin/env node
/**
 * Frontend Error Scanner
 * Scans TypeScript/React files for errors and issues.
 */

const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");

// Colors for terminal output
const colors = {
  reset: "\x1b[0m",
  red: "\x1b[31m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  blue: "\x1b[34m",
};

function printHeader(text) {
  console.log(`\n${colors.blue}${"=".repeat(60)}`);
  console.log(text);
  console.log(`${"=".repeat(60)}${colors.reset}\n`);
}

function printSuccess(text) {
  console.log(`${colors.green}✓ ${text}${colors.reset}`);
}

function printError(text) {
  console.log(`${colors.red}✗ ${text}${colors.reset}`);
}

function printWarning(text) {
  console.log(`${colors.yellow}⚠ ${text}${colors.reset}`);
}

function runCommand(command, description) {
  printHeader(description);
  try {
    const output = execSync(command, { encoding: "utf-8", stdio: "pipe" });
    if (output.trim()) {
      console.log(output);
    }
    printSuccess(`${description} completed successfully!`);
    return { success: true, output };
  } catch (error) {
    printError(`${description} failed!`);
    if (error.stdout) console.log(error.stdout);
    if (error.stderr) console.log(error.stderr);
    return { success: false, error: error.message };
  }
}

function checkTypeScript() {
  printHeader("Checking TypeScript Compilation");
  try {
    execSync("npx tsc --noEmit", { encoding: "utf-8", stdio: "pipe" });
    printSuccess("No TypeScript errors found!");
    return { success: true, errors: 0 };
  } catch (error) {
    const output = error.stdout || error.stderr || "";
    const errorCount = (output.match(/error TS/g) || []).length;
    printError(`Found ${errorCount} TypeScript errors:`);
    console.log(output);
    return { success: false, errors: errorCount };
  }
}

function runESLint() {
  printHeader("Running ESLint");
  try {
    execSync("npm run lint", { encoding: "utf-8", stdio: "pipe" });
    printSuccess("No ESLint errors found!");
    return { success: true, errors: 0 };
  } catch (error) {
    const output = error.stdout || error.stderr || "";
    const errorLines = output
      .split("\n")
      .filter((line) => line.includes("error") || line.includes("warning"));
    printWarning(`ESLint found ${errorLines.length} issues:`);
    console.log(output);
    return { success: false, errors: errorLines.length };
  }
}

function checkBuildSize() {
  printHeader("Checking Build Configuration");

  const nextConfigPath = path.join(process.cwd(), "next.config.js");
  const nextConfigMjsPath = path.join(process.cwd(), "next.config.mjs");

  if (fs.existsSync(nextConfigPath) || fs.existsSync(nextConfigMjsPath)) {
    printSuccess("Next.js config found");
    return { success: true };
  } else {
    printWarning("Next.js config not found");
    return { success: false };
  }
}

function checkDependencies() {
  printHeader("Checking Dependencies");
  try {
    execSync("npm ls --depth=0", { encoding: "utf-8", stdio: "pipe" });
    printSuccess("All dependencies are installed correctly!");
    return { success: true };
  } catch (error) {
    printWarning("Some dependency issues found:");
    console.log(error.stdout || error.message);
    return { success: false };
  }
}

function scanUnusedDependencies() {
  printHeader("Scanning for Unused Dependencies");
  try {
    // Check if depcheck is installed
    execSync("npx depcheck --version", { stdio: "pipe" });
    const output = execSync("npx depcheck --json", { encoding: "utf-8" });
    const result = JSON.parse(output);

    if (result.dependencies && result.dependencies.length > 0) {
      printWarning(`Found ${result.dependencies.length} unused dependencies:`);
      result.dependencies.forEach((dep) => console.log(`  - ${dep}`));
      return { success: false, unused: result.dependencies.length };
    } else {
      printSuccess("No unused dependencies found!");
      return { success: true, unused: 0 };
    }
  } catch (error) {
    printWarning(
      "depcheck not available. Install with: npm install -g depcheck",
    );
    return { success: true, unused: 0 };
  }
}

function generateReport(results) {
  printHeader("SCAN SUMMARY");

  console.log(
    `TypeScript: ${
      results.typescript.success
        ? "✓ PASS"
        : `✗ FAIL (${results.typescript.errors} errors)`
    }`,
  );
  console.log(
    `ESLint: ${
      results.eslint.success
        ? "✓ PASS"
        : `✗ FAIL (${results.eslint.errors} issues)`
    }`,
  );
  console.log(
    `Build Config: ${results.buildConfig.success ? "✓ PASS" : "✗ FAIL"}`,
  );
  console.log(
    `Dependencies: ${results.dependencies.success ? "✓ PASS" : "✗ FAIL"}`,
  );
  console.log(
    `Unused Deps: ${
      results.unusedDeps.unused === 0
        ? "✓ PASS"
        : `⚠ ${results.unusedDeps.unused} found`
    }`,
  );

  const totalErrors =
    (results.typescript.errors || 0) + (results.eslint.errors || 0);

  console.log(`\n${colors.blue}Total Errors: ${totalErrors}${colors.reset}`);

  if (
    totalErrors === 0 &&
    results.buildConfig.success &&
    results.dependencies.success
  ) {
    console.log(
      `\n${colors.green}🎉 All checks passed! Your code looks good!${colors.reset}`,
    );
  } else {
    console.log(
      `\n${colors.yellow}⚠️  Please fix the issues above before deploying.${colors.reset}`,
    );
  }

  return totalErrors === 0;
}

function main() {
  console.log(`${colors.blue}Frontend Error Scanner${colors.reset}`);
  console.log(`Scanning directory: ${process.cwd()}\n`);

  const results = {
    typescript: checkTypeScript(),
    eslint: runESLint(),
    buildConfig: checkBuildSize(),
    dependencies: checkDependencies(),
    unusedDeps: scanUnusedDependencies(),
  };

  const success = generateReport(results);
  process.exit(success ? 0 : 1);
}

main();
